src/uu/dmesg/src/time_formatter.rs

```diff
@@ -0,0 +1,166 @@
+use chrono::{DateTime, FixedOffset, TimeDelta};
+#[cfg(feature = "fixed-boot-time")]
+use chrono::{NaiveDate, NaiveTime};
+use std::sync::OnceLock;
+
+pub fn raw(timestamp_us: i64) -> String {
+    let seconds = timestamp_us / 1000000;
+    let sub_seconds = timestamp_us % 1000000;
+    format!("{:>5}.{:0>6}", seconds, sub_seconds)
+}
+
+pub fn ctime(timestamp_us: i64) -> String {
+    let date_time = boot_time()
+        .checked_add_signed(TimeDelta::microseconds(timestamp_us))
+        .unwrap();
+    date_time.format("%a %b %d %H:%M:%S %Y").to_string()
+}
+
+pub fn iso(timestamp_us: i64) -> String {
+    let date_time = boot_time()
+        .checked_add_signed(TimeDelta::microseconds(timestamp_us))
+        .unwrap();
+    date_time.format("%Y-%m-%dT%H:%M:%S,%6f%:z").to_string()
+}
+
+pub struct ReltimeFormatter {
+    state: State,
+    prev_timestamp_us: i64,
+    previous_unix_timestamp: i64,
+}
+
+pub struct DeltaFormatter {
+    state: State,
+    prev_timestamp_us: i64,
+}
+
+pub enum State {
+    Initial,
+    AfterBoot,
+    Delta,
+}
+
+impl ReltimeFormatter {
+    pub fn new() -> Self {
+        ReltimeFormatter {
+            state: State::Initial,
+            prev_timestamp_us: 0,
+            previous_unix_timestamp: 0,
+        }
+    }
+
+    pub fn format(&mut self, timestamp_us: i64) -> String {
+        let date_time = boot_time()
+            .checked_add_signed(TimeDelta::microseconds(timestamp_us))
+            .unwrap();
+        let unix_timestamp = date_time.timestamp();
+        let minute_changes = (unix_timestamp / 60) != (self.previous_unix_timestamp / 60);
+        let format_res = match self.state {
+            State::Initial => date_time.format("%b%d %H:%M").to_string(),
+            _ if minute_changes => date_time.format("%b%d %H:%M").to_string(),
+            State::AfterBoot => Self::delta(0),
+            State::Delta => Self::delta(timestamp_us - self.prev_timestamp_us),
+        };
+        self.prev_timestamp_us = timestamp_us;
+        self.previous_unix_timestamp = unix_timestamp;
+        self.state = match self.state {
+            State::Initial if timestamp_us == 0 => State::AfterBoot,
+            _ => State::Delta,
+        };
+        format_res
+    }
+
+    fn delta(delta_us: i64) -> String {
+        let seconds = i64::abs(delta_us / 1000000);
+        let sub_seconds = i64::abs(delta_us % 1000000);
+        let sign = if delta_us >= 0 { '+' } else { '-' };
+        let mut res = format!("{}.{:0>6}", seconds, sub_seconds);
+        res.insert(0, sign);
```

Comment:
What's the reason for using `insert` to add the sign instead of doing it on line 77?

Reply:
Ah yes, I could've added it on line 77 for this case. I changed it in commit e965074.

Just curious, your comment regarding this is concerning readability, correct? Or are you also concerned about the string processing performance?

Reply:
Yes, it is about readability. I don't know whether there is much of a difference in string processing performance.